# Capítulo 2. Funciones y Condicionales

En este capítulo, introduciremos los conceptos de funciones y sentencias condicionales. Las funciones nos permiten encapsular bloques de código reutilizable, haciendo programas más organizados y eficientes. Las sentencias condicionales nos permiten ejecutar diferentes partes de nuestro código basándonos en si ciertas condiciones son verdaderas o falsas, lo que añade lógica y capacidad de toma de decisiones a los programas.

### 2.1 Definición de Funciones en Python (def)

Una función, es un bloque de código con un nombre asociado que realiza una tarea específica. Las funciones nos permiten dividir programas complejos en partes más pequeñas y manejables, promueven la reutilización de código y mejoran la legibilidad. En Python, definimos una función utilizando la palabra clave “def”, seguida del nombre de la función, paréntesis (), y dos puntos “:”. El bloque de código que pertenece a la función debe estar indentado, y se le conoce como el cuerpo de la función.

In [1]:

def saludar():
    """Función para saludar"""
    # Para ejecutar la función, la llamamos por su nombre seguido de paréntesis: saludar()
    print("¡Hola desde una función!")

saludar()

¡Hola desde una función!


#### 2.1.1 Argumentos y Parámetros

Las funciones pueden recibir información del exterior a través de argumentos. Cuando definimos una función, especificamos parámetros, que son variables que recibirán los valores de los argumentos cuando la función sea llamada.

In [2]:
def calcular_impuesto(ingreso, tasa_impositiva):
    """  """
    impuesto = ingreso * tasa_impositiva
    print("El impuesto a pagar es:", impuesto)


# Al llamar la función, proporcionamos los argumentos:
calcular_impuesto(50000, 0.20)
calcular_impuesto(75000, 0.25)

El impuesto a pagar es: 10000.0
El impuesto a pagar es: 18750.0


En este ejemplo, ingreso y tasa_impositiva son los parámetros de la función, mientras que 50000 y 0.20, y 75000 y 0.25 son los argumentos en las diferentes llamadas a la función.

#### 2.1.2 Valores de Retorno (return)

Además de realizar acciones (como imprimir en pantalla), las funciones a menudo necesitan devolver un resultado. Utilizamos la palabra clave return para especificar el valor que la función debe devolver. Una vez que se ejecuta una sentencia return, la función termina.

In [3]:
def calcular_pib(consumo, inversion, gasto_publico, exportaciones, importaciones):
    """  """
    pib = consumo + inversion + gasto_publico + exportaciones - importaciones
    return pib

pib_año_2023 = calcular_pib(1500, 500, 700, 600, 400)
print(pib_año_2023)

2900


La función calcular_pib toma varios argumentos y devuelve el valor calculado del PIB. Este valor puede luego ser utilizado en otras partes de nuestro programa.

### 2.2 Sentencias Condicionales (if, elif, else)

Las sentencias condicionales nos permiten ejecutar diferentes bloques de código dependiendo de si una o más condiciones son verdaderas o falsas. Python utiliza las palabras clave “if”, “elif” (abreviatura de "else if"), y “else” para definir estas condiciones. Estas expresiones a menudo involucran operadores de comparación como los vistos en el apartado “1.2.4 Operadores de comparación” (== igual a, != no igual a, < menor que, > mayor que, <= menor o igual que, >= mayor o igual que),

Una forma de interpretar en español estas condiciones es:
- if: si sucede esta condición  
- elif: si sucede esta otra condición
- else: cualquier condición diferente de las condiciones anteriores

Al igual que las funciones, las sentencias condicionales hacen uso de la indentación para especificar las secciones de código que forman parte de la sentencia condicional, la estructura de una sentencia condicional sigue la forma del siguiente ejemplo, donde pregunta si 5 es igual a 5, como esto es verdadero (True), entonces python ejecuta el bloque de código identado.

In [4]:
if (5 == 5):
    print("Evaluación de la condición:", 5 == 5)

Evaluación de la condición: True


Si la sentencia condicional es falsa (False), python no ingresa al bloque de código identado, el siguiente ejemplo muestra el comportamiento cuando la sentencia condicional es falsa (False) y se cumple la condición “else”.

In [5]:
if 5 == 4:
    print("Esta línea de código no se ejecuta")
else:
    print("Evaluación de la condición:", 5 == 4)

Evaluación de la condición: False


Como la primera condición “if” es falsa (False) Python pregunta por la siguiente condición, en este caso “else”, que como vimos anteriormente podemos interpretarla como “aquello diferente de las otras condiciones”: en este caso sería verdadero (True), ya que verdadero es lo opuesto de falso en la primera condición “if”, con lo cual, se ejecuta el bloque de código identado posterior al “else”.  
Resulta importante mencionar que, en el ejemplo anterior, al evaluar la primera condición del “if”, no se utilizaron paréntesis para la evaluación de la condición “if 5 == 4”, a diferencia del ejemplo “if (5 == 5)”. Python permite evaluar condiciones sin el uso de paréntesis, sin embargo, se recomienda utilizar paréntesis al evaluar condiciones, ya que como veremos más adelante, existen evaluaciones más complejas donde el uso de paréntesis resulta necesario para la ejecución correcta del código.
Por último, el siguiente ejemplo muestra las tres condiciones “if”, “elif” y “else”, es importante mencionar que después de la condición “if” se pueden agregar cualquier cantidad de condiciones “elif” que se consideren necesarias.

In [6]:
x= 15

if (x > 20): # se lee como: si la variable “x” es mayor que 20
    print("La variable", x, " es mayor que 20")
elif (x< 15): # se lee como si la variable “x” es menor que 15
    print("La variable", x, " es menor que 15")
else: # se lee como: aquello diferente a las otras condiciones
    print("La variable", x, " es mayor o igual a 15 y menor o igual a 20")

La variable 15  es mayor o igual a 15 y menor o igual a 20


#### 2.2.1 Operadores lógicos (and, or, not) con sentencias condicionales

Los operadores lógicos permiten combinar o modificar expresiones condicionales para tomar decisiones más complejas. En Python, los operadores lógicos básicos son “and” (y), “or” (o) y “not” (no).
- Operador “and” (y lógico): evalúa si todas las condiciones involucradas son verdaderas, devuelve “True” solo si ambos operandos son “True”. Si alguno es falso, el resultado es “False”. Python evalúa de izquierda a derecha y se detiene en el primer “False”.
- Operador “or” (o lógico): evalúa si al menos una de las condiciones es verdadera, devuelve “True” si alguno de los operandos es “True”. Solo retorna False si ambos son False. Python se detiene en el primer “True”.
- Operador “not” (negación): invierte el valor booleano de una condición, y convierte “True” en “False” y viceversa.

In [7]:
tasa_desempleo = 0.07
inflacion = 0.03

if (tasa_desempleo > 0.08) and (inflacion > 0.02):
    print("La economía enfrenta estanflación.")
elif (tasa_desempleo > 0.08):
    print("Hay un alto nivel de desempleo.")
elif (inflacion > 0.02):
    print("Hay una alta inflación.")
else:
    print("La economía parece estable.")

Hay una alta inflación.


En el ejemplo anterior, la condición en la sentencia “if” se evalúa primero. Si es verdadera, se ejecuta el bloque de código indentado debajo de ella. Si es falsa, se evalúa la condición en la siguiente sentencia elif, y así sucesivamente. Si ninguna de las condiciones if o elif es verdadera, se ejecuta el bloque de código bajo la sentencia “else” (si está presente). Nótese que las condiciones “elif” y “else” no están identadas, están al mismo nivel que la condición inicial “if”; al igual que en las funciones, es importante mencionar que después de la condición se deben colocar dos puntos “:”, esto indica el final de la condición, las siguientes líneas de código que se ejecutan deben ir indentadas para establecer que ese código pertenece a esa condición. Así mismo, el código no indentado es otra condición o puede ser código que no depende de ninguna de estas condiciones.

A continuación, se muestran otro ejemplo haciendo uso del operador “or”.

In [8]:
# Datos de un solicitante
puntaje_credito = 650
tiene_historial_impagos = True # El cliente tuvo un impago en el pasado

# La primera condición es Falsa (650 no es < 600)

if (puntaje_credito < 600) or (tiene_historial_impagos == True):
    print("La solicitud requiere una revisión adicional por riesgo.")
else:
    print("La solicitud pasa el filtro de riesgo inicial.")

La solicitud requiere una revisión adicional por riesgo.


En el código anterior, Python evalúa cada lado de la expresión:


1.  **Lado Izquierdo:** “puntaje_credito < 600” se convierte en “650 < 600”, lo cual es **False.**
2.  **Lado Derecho:** “tiene_historial_impagos == True” se convierte en “True == True”, lo cual es **True**

Como el “or” solo necesita que al menos una de las condiciones sea True para que toda la expresión sea True, la condición general se cumple y el programa ejecuta el código para marcar la solicitud como riesgosa.

Por último, el siguiente ejemplo muestra la condición “not”.


In [9]:
# Datos de un visitante del sitio web
es_miembro_registrado = False # Este usuario es nuevo

# 'not' invierte el valor booleano de la variable
if not es_miembro_registrado:
    print("¡Bienvenido! Hemos aplicado un descuento especial!.")
else:
    print("Gracias por ser parte de nuestra comunidad.")

¡Bienvenido! Hemos aplicado un descuento especial!.


En el código anterior, la línea clave es: “if not es_miembro_registrado”. El operador “not” actúa como un **inversor**. Simplemente toma el valor booleano que le sigue y lo voltea al opuesto.


1.  **Valor Original:** La variable es_miembro_registrado contiene el valor **False.**
2.  **Aplicando not:** La expresión se convierte en **not False.**
1.   **Resultado Final:** Python evalúa not False como **True.**


Como la condición final del “if” es “True”, el programa ejecuta el bloque de código que da la bienvenida al nuevo cliente.


#### 2.3 Funciones nativas y funciones en NumPy

Ahora que conocemos lo que es una función y las sentencian condicionales, podemos hacer uso de estos elementos y aplicarlos al análisis estadístico descriptivo. Como parte final de este capítulo, mostraremos la forma de programar funciones para el cálculo de la media, sumatorias y la mediana, así como las herramientas que Python ofrece para estos cálculos, desde funciones nativas como “sum()”, “len()”, “min()”, “max()”, etc.., hasta librerías especializadas como NumPy con operaciones vectorizadas.: exploraremos ambos enfoques, pero antes mostraremos algunas funciones nativas básicas de Python:

- sum(): Suma los elementos de un iterable.
- len(): Cuenta elementos en un iterable.
- min() / max(): Encuentra el valor mínimo/máximo.
- round(): Redondea a un número de decimales.
- sorted(): Ordena un iterable.


In [10]:
ventas = [5000, 6000, 4500]  # Datos originales de ventas

# Cálculos estadísticos (incluyendo sorted)
total = sum(ventas)
promedio = round(total / len(ventas), 2)
max_venta = max(ventas)
min_venta = min(ventas)
ventas_ordenadas = sorted(ventas)  # Nueva lista ordenada de menor a mayor

# Resultados
print(f"Total: ${total}")
print(f"Promedio: ${promedio}")
print(f"Venta máxima: ${max_venta}")
print(f"Venta mínima: ${min_venta}")
print(f"Ventas ordenadas: {ventas_ordenadas}")

Total: $15500
Promedio: $5166.67
Venta máxima: $6000
Venta mínima: $4500
Ventas ordenadas: [4500, 5000, 6000]


##### Caluclar suma

Tres formas diferentes de calcular la suma de elementos, mostrando las diferencias entre métodos nativos de Python y operaciones vectorizadas de NumPy.

In [11]:
import numpy as np

ingresos = [2.5, 3.0, 4.5, 1.8, 2.2, 3.8, 4.0] # lista de python
ingresos_np = np.array(ingresos) # array NumPy

def suma(x):
    """Calcula la suma usando producto punto con vector de unos."""
    unos = np.ones(len(x))  # Vector de unos del mismo tamaño
    return np.dot(x, unos) # función producto punto


print("Función suma Pyhton:", sum(ingresos))
print("Función suma numpy:", np.sum(ingresos_np))
print("Función suma manual:", suma(ingresos_np))

Función suma Pyhton: 21.8
Función suma numpy: 21.8
Función suma manual: 21.8


El código presentado calcula la suma de una lista de ingresos utilizando tres métodos diferentes: la función nativa sum() de Python, la función np.sum() de NumPy y una implementación manual basada en el producto punto (suma()). El propósito es ilustrar las diferencias en términos de eficiencia, legibilidad y aplicabilidad, especialmente en el contexto del análisis económico con Python.

- La función nativa sum() de Python opera sobre una lista de Python (en este caso, ingresos). Es simple y adecuada para conjuntos de datos pequeños, pero su rendimiento disminuye significativamente con grandes volúmenes de datos debido a la falta de optimización para cálculos numéricos.
- np.sum() de NumPy está diseñada para cálculos numéricos eficientes. Esta función es mucho más rápida que sum() en datasets grandes, gracias a su implementación y optimizaciones para operaciones vectorizadas.
- Una implementación manual (suma()) que usa producto punto con un vector de unos, la cual, aunque ilustra el fundamento algebraico (equivalente a sumar elementos), es innecesariamente compleja y menos eficiente en la práctica, ya que no aprovecha las optimizaciones de NumPy

Si bien todos los métodos devuelven el mismo resultado (21.8), la elección depende del contexto: simplicidad, eficiencia o claridad conceptual, en este caso estaremos utilizando las funciones de NumPy por su eficiencia y aplicabilidad a datos económicos.

En el siguiente ejemplo calcularemos la media aritmética y mostraremos que podemos llamar a funciones dentro de otras funciones, utilizando la funcion “suma()” que definimos de forma “manual”.

##### __Calcular la media__

In [12]:
import numpy as np

ingresos = [2.5, 3.0, 4.5, 1.8, 2.2, 3.8, 4.0]
ingresos_np = np.array(ingresos)

def media(datos):
    """Calcula la media aritmética."""
    return suma(datos) / len(datos)

media_manual = media(ingresos)
media_np = np.mean(ingresos_np)

print("Media manual:", media_manual, "Media numpy:", media_np, sep="\n")

Media manual:
3.1142857142857143
Media numpy:
3.1142857142857143


##### __Calcular la mediana__

In [13]:
import numpy as np

def mediana(datos):
    """Calcula la mediana (valor central) sin usar np.median."""
    datos_ordenados = sorted(datos)
    n = len(datos_ordenados)
    mitad = n // 2
    if n % 2 == 1:
        return datos_ordenados[mitad]
    else:
        return (datos_ordenados[mitad - 1] + datos_ordenados[mitad]) / 2

media_manual = media(ingresos)
media_np = np.mean(ingresos_np)

print(media_manual)
print(media_np)

3.1142857142857143
3.1142857142857143


La razón de elegir la sumatoria, media y mediana en esta seccion, es para afianzar los conocimientos hasta ahora definidos en capitulos anteriores, así como mostrar las funciones de python para estos cálculos (nativas y de NumPy), mostrando las diferencias y los casos de uso de cada uno de estos elementos, en los siguientes capítulos mostraremos a manera de ejemplo la forma de programar otros métodos de análisis estadístico descriptivo, así como operaciones más complejas en el ámbito de la economía.